File: ios/Sources/OberonTUI/TerminalState.swift
// TerminalState — observable model behind the SwiftUI TerminalView.
//
// The Oberon-side TUI calls (Init / Clear / MoveTo / Write / Flush)
// route through TerminalShim.swift's @_cdecl entry points and mutate
// this shared instance; SwiftUI observes the @Published properties
// and redraws. Mirrors the role of the web tui-shim's local `cells`
// / `curRow` / `curCol` state — single-host invariant: there is one
// terminal per process.

import SwiftUI

/// Display attribute bits — must match TUI.Mod's AttrReverse / Bold /
/// Italic / Underline.
public struct CellAttr: OptionSet, Sendable {
    public let rawValue: Int32
    public init(rawValue: Int32) { self.rawValue = rawValue }
    public static let reverse   = CellAttr(rawValue: 1)
    public static let bold      = CellAttr(rawValue: 2)
    public static let italic    = CellAttr(rawValue: 4)
    public static let underline = CellAttr(rawValue: 8)
}

/// One screen cell. `fg`/`bg` are indices into the 3270 palette
/// (0..7) or -1 for "terminal default".
public struct Cell: Sendable, Equatable {
    public var ch:   Character
    public var attr: CellAttr
    public var fg:   Int8       // -1 = default
    public var bg:   Int8

    public static let blank = Cell(ch: " ", attr: [], fg: -1, bg: -1)
}

/// Shared terminal state. The shim mutates this from the main thread
/// (Oberon callbacks run on the same thread that dispatches keys, i.e.
/// the main thread under SwiftUI). The @Published wrappers trigger
/// SwiftUI redraws when `flush` advances `frame`.
@MainActor
public final class TerminalState: ObservableObject {
    public static let shared = TerminalState()

    /// Grid dimensions in cells. Updated by TerminalView's geometry
    /// reader before any Oberon code runs and on every size change;
    /// the ios_tui_rows / ios_tui_cols externs read these.
    public var rows: Int = 24
    public var cols: Int = 80

    /// Pending state being written by Oberon. The view reads from
    /// `displayed` (committed at flush); this two-buffer scheme
    /// avoids partial-frame flicker during a multi-call redraw.
    public var pending:   [[Cell]] = []
    @Published public var displayed: [[Cell]] = []

    public var curCol = 0
    public var curRow = 0
    public var curAttr: CellAttr = []
    public var curFg:   Int8 = -1
    public var curBg:   Int8 = -1
    @Published public var cursorVisible = true

    /// Bumped on every flush so views that key off it can force
    /// updates without diffing the full grid.
    @Published public var frame: Int = 0

    private init() {
        reset()
    }

    public func reset() {
        pending   = Self.makeGrid(rows: rows, cols: cols)
        displayed = pending
        curCol = 0; curRow = 0
        curAttr = []; curFg = -1; curBg = -1
    }

    /// Resize the grid in-place. Existing cells survive where they
    /// still fit; new rows / columns blank-fill. Critical for the
    /// startup race: SwiftUI's GeometryReader fires resize once or
    /// twice during initial layout, and we don't want to wipe out
    /// whatever the Oberon side may have already painted.
    public func resize(rows: Int, cols: Int) {
        let newRows = max(1, rows)
        let newCols = max(1, cols)
        guard newRows != self.rows || newCols != self.cols else { return }

        if newRows > pending.count {
            let extra = Array(repeating: Array(repeating: Cell.blank, count: newCols),
                              count: newRows - pending.count)
            pending.append(contentsOf: extra)
        } else if newRows < pending.count {
            pending = Array(pending.prefix(newRows))
        }
        for r in 0..<pending.count {
            if pending[r].count < newCols {
                pending[r].append(contentsOf:
                    Array(repeating: Cell.blank, count: newCols - pending[r].count))
            } else if pending[r].count > newCols {
                pending[r] = Array(pending[r].prefix(newCols))
            }
        }

        self.rows = newRows
        self.cols = newCols
        displayed = pending
        if curRow >= self.rows { curRow = self.rows - 1 }
        if curCol >= self.cols { curCol = self.cols - 1 }
    }

    public func putChar(_ ch: Character) {
        guard curRow >= 0, curRow < rows, curCol >= 0, curCol < cols else { return }
        pending[curRow][curCol] = Cell(ch: ch, attr: curAttr, fg: curFg, bg: curBg)
        curCol += 1
        if curCol >= cols { curCol = cols - 1 }
    }

    public func clear() {
        for r in 0..<rows {
            for c in 0..<cols {
                pending[r][c] = .blank
            }
        }
        curCol = 0; curRow = 0
    }

    public func clearLine() {
        guard curRow >= 0, curRow < rows else { return }
        for c in 0..<cols {
            pending[curRow][c] = .blank
        }
    }

    public func moveTo(col: Int, row: Int) {
        curCol = max(0, min(col, cols - 1))
        curRow = max(0, min(row, rows - 1))
    }

    public func flush() {
        displayed = pending
        frame &+= 1
    }

    private static func makeGrid(rows: Int, cols: Int) -> [[Cell]] {
        Array(repeating: Array(repeating: Cell.blank, count: cols), count: rows)
    }
}

File: android/src/main/kotlin/com/oneav/oberon/tui/TerminalState.kt
// TerminalState — observable model behind the Compose TerminalView.
//
// Mirrors the iOS TerminalState: two-buffer cell grid, `putChar`
// mutates `pending`, `flush` commits to `displayed`, Compose observes
// `frame`. All access is expected on the UI thread; JNI upcalls
// from Oberon C code run there so there's no explicit locking.

package com.oneav.oberon.tui

import androidx.compose.runtime.mutableStateOf

/** Display attribute bits — must match TUI.Mod's AttrReverse /
 *  Bold / Italic / Underline. */
object CellAttr {
    const val REVERSE   = 1
    const val BOLD      = 2
    const val ITALIC    = 4
    const val UNDERLINE = 8
}

/** One screen cell. `fg`/`bg` are indices into the 3270 palette
 *  (0..7) or -1 for "terminal default". */
data class Cell(
    val ch:   Char,
    val attr: Int,
    val fg:   Int,
    val bg:   Int,
) {
    companion object {
        val BLANK = Cell(ch = ' ', attr = 0, fg = -1, bg = -1)
    }
}

object TerminalState {
    /** Grid dimensions in cells. Updated by TerminalView's geometry
     *  observer before Oberon reads via tuiRows / tuiCols. */
    var rows: Int = 24
        private set
    var cols: Int = 80
        private set

    /** Pending state being written by Oberon; committed to
     *  `displayed` on flush. The two-buffer scheme avoids
     *  partial-frame flicker during a multi-call redraw. */
    private var pending: Array<Array<Cell>> = makeGrid(rows, cols)

    /** What Compose renders. Compose diffs the two-dimensional
     *  Array reference; we allocate a fresh grid on flush so the
     *  state observer definitely fires. */
    val displayed = mutableStateOf(pending)

    var curCol: Int = 0
        private set
    var curRow: Int = 0
        private set
    private var curAttr: Int = 0
    private var curFg:   Int = -1
    private var curBg:   Int = -1
    val cursorVisible = mutableStateOf(true)

    /** Bumped on every flush so views that key off it can force
     *  redraws without diffing the full grid. */
    val frame = mutableStateOf(0)

    init { reset() }

    fun reset() {
        pending = makeGrid(rows, cols)
        displayed.value = pending
        curCol = 0; curRow = 0
        curAttr = 0; curFg = -1; curBg = -1
        cursorVisible.value = true
    }

    /** Grow / shrink in place — never wipes, so Compose's initial
     *  layout passes don't blank what Oberon may have already
     *  painted. Matches the iOS resize semantics. */
    fun resize(newRows: Int, newCols: Int) {
        val r = maxOf(1, newRows)
        val c = maxOf(1, newCols)
        if (r == rows && c == cols) return
        val next = Array(r) { row ->
            Array(c) { col ->
                if (row < pending.size && col < pending[row].size)
                    pending[row][col]
                else
                    Cell.BLANK
            }
        }
        pending = next
        rows = r; cols = c
        displayed.value = pending
        if (curRow >= rows) curRow = rows - 1
        if (curCol >= cols) curCol = cols - 1
    }

    fun putChar(ch: Char) {
        if (curRow !in 0 until rows || curCol !in 0 until cols) return
        pending[curRow][curCol] = Cell(ch, curAttr, curFg, curBg)
        curCol++
        if (curCol >= cols) curCol = cols - 1
    }

    fun clear() {
        for (r in 0 until rows) for (c in 0 until cols) {
            pending[r][c] = Cell.BLANK
        }
        curCol = 0; curRow = 0
    }

    fun clearLine() {
        if (curRow !in 0 until rows) return
        for (c in 0 until cols) pending[curRow][c] = Cell.BLANK
    }

    fun moveTo(col: Int, row: Int) {
        curCol = col.coerceIn(0, cols - 1)
        curRow = row.coerceIn(0, rows - 1)
    }

    fun setAttr(a: Int) { curAttr = a }
    fun setFg(c: Int)   { curFg = if (c in -1..7) c else -1 }
    fun setBg(c: Int)   { curBg = if (c in -1..7) c else -1 }
    fun setCursorVisible(v: Boolean) { cursorVisible.value = v }

    fun flush() {
        // Allocate a fresh outer array so Compose's reference-based
        // observation triggers. Row references are shared with
        // pending — the next paint operation mutates in place and
        // the next flush produces a fresh outer snapshot.
        val snapshot = Array(rows) { pending[it] }
        displayed.value = snapshot
        frame.value += 1
    }

    private fun makeGrid(r: Int, c: Int): Array<Array<Cell>> =
        Array(r) { Array(c) { Cell.BLANK } }
}

/** Polling-path key queue. Event-driven dispatch (via
 *  OberonBridge.ocDispatchKey) bypasses it; if the program
 *  registers a key handler, this stays empty. */
object KeyQueue {
    private val queue = ArrayDeque<Int>()
    fun push(code: Int) { queue.addLast(code) }
    fun pop(): Int? = if (queue.isEmpty()) null else queue.removeFirst()
}
